### PR TITLE
HPCC-32306 Optimize REGEXFIND(,n) to a fixed-sized target

### PR DESCRIPTION
**Pull request by dcamper**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
Looks good from initial scan.  I will run regression test and changes in the output

File: testing/regress/ecl/regex_find_fixed.ecl [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+STRING first_word_0_ps := '^\\w+' : STORED('first_word_0_ps');
+UNICODE first_word_0_pu := u'^\\p{L}+' : STORED('first_word_0_pu');
+
+first_word_0 := PROJECT
+    (
+        NOFOLD(inDS),
+        TRANSFORM
+            (
+                ResLayout,
+                SELF.unbounded_s := NOFOLD(REGEXFIND(first_word_0_ps, (STRING)LEFT.s, 0)),
```
Comment (ghalliday) [inline testing/regress/ecl/regex_find_fixed.ecl:46]: This test case doesn't actually change with your code enabled.  The same search is assigned to variable and fixed length targets, which means that the variable length result is reused.

**Comment by ghalliday** [comment]
Interestingly this has very little effect on most of the test cases.  I think it is because the result of the REGEXFIND is also used later on in the transform - similar to your example code.  I only saw 3 examples that shrank by more than 0.1%, despite some examples having > 70,000 regexfinds.

**Comment by dcamper** [comment]
The question then becomes, is the minor increase in complexity worth the minor increase in efficiency? It might be a total wash, in which case we may want to err on the side of "less is better." Thoughts?

File: testing/regress/ecl/regex_find_fixed.ecl (hunk 1 above)
Comment (dcamper) [reply testing/regress/ecl/regex_find_fixed.ecl]: I pushed a change where all of the unbounded assignments were removed. That was the only way I could find to (easily) get the code generator to use the fixed-length versions.

**Comment by dcamper** [comment]
@ghalliday Squashed and rebased to master branch.

**Pull request merged**